### PR TITLE
Functions concept

### LINKED ISSUE #1642: [Sept Sprint] New exercise teaching functions
This is part of our September Spring to get the Go Track in shape. Please read the [Overview issue](https://github.com/exercism/go/issues/1610) for more details. **This exercise is unclaimed. To claim it, please comment below.**

---

**Before building this exercise (or exercises), please read up on [the documentation](https://exercism.org/docs/building/tracks/concept-exercises).**

We need an exercise for the `functions` concept, which should teach about:

- Arguments (incl. syntax for arguments of the same type)
- Return values
- Named return values
- Naked return
- pass by value but the value can be a pointer
- Can also include: Anonymous functions/callbacks

It should **not** teach (these will be covered later):

- Variadic functions (that should be introduced with rest-and-spread)

It is worth to look if it is possible to port this exercise from the JS [lasagna master exercise](https://github.com/exercism/javascript/tree/main/exercises/concept/lasagna-master/.docs).

### PR DESCRIPTION
**Pull request by bobtfish**
This is my first attempt at something as ambitious as a full exercise and a concept, so any/all feedback and tips very gratefully received.

Specific things that I'm unsure about / have questions on:
  * Fitting this concept/exercise into the existing hierarchy - I've done this somewhat, but if I add dependencies for everything that's relevant then I create cycles. Feedback welcomes from someone who has a more holistic view of the progression of concepts they think is right for the Go track.
  * The concept introduction and exercise introduction are nearly identical. Is this OK/appropriate? (Should I just be using .docs/introduction.md.tpl if so?)
  * .meta/design.md - this is my first attempt at one of these, I've read a few others but it's largely cargo cult.
  * Should my exercise include `forked_from`? It's not clear to me from the docs what the contents of this field would be? (I.e. is it just same track exercises, or can you reference cross tracks etc)

Fixes #1642

### CODE REVIEW

File: exercises/concept/lasagna-master/.meta/config.json
Comment (ErikSchierboom) [inline exercises/concept/lasagna-master/.meta/config.json:11]: You don't need the source if you specify the forked_from field (which is appropriate here)

```suggestion
  "forked_from": ["javascript/lasagna-master"]
```

File: exercises/concept/lasagna-master/.meta/exemplar.go
Comment (ErikSchierboom) [inline exercises/concept/lasagna-master/.meta/exemplar.go:4]: Should these functions start with a capital letter? I've seen that used in other exercises.

File: exercises/concept/lasagna-master/lasagna-master.go
Comment (ErikSchierboom) [inline exercises/concept/lasagna-master/lasagna-master.go:11]: I'd recommend doing the same here as in the [lasagna-exercise](https://github.com/exercism/go/blob/main/exercises/concept/lasagna/lasagna.go#L5), where there are TODOs for individual functions.

```suggestion
// TODO: define the 'CookingStatus()' function
```

File: concepts/functions/about.md
Comment (ErikSchierboom) [inline concepts/functions/about.md:116]: There will be a separate exercise on pointers: https://github.com/exercism/go/issues/1643

```suggestion
```

File: concepts/functions/about.md
Comment (ErikSchierboom) [inline concepts/functions/about.md:31]: Nicely explained!

File: concepts/functions/about.md
Comment (ErikSchierboom) [inline concepts/functions/about.md:5]: I'm not a fan of referring to other concepts like this upfront. Maybe change these to be [concept widget links](https://exercism.org/docs/building/markdown/widgets) and move this text somewhat later in this document?

File: concepts/functions/about.md
Comment (ErikSchierboom) [inline concepts/functions/about.md:156]: Variadic functions are out of scope for this exercise (see the [issue description](https://github.com/exercism/go/issues/1642#issue-999186434)). Could you remove this section?

File: concepts/functions/about.md
Comment (ErikSchierboom) [inline concepts/functions/about.md:119]: @junedev Do we want to include public/private here?

File: concepts/functions/introduction.md
Comment (ErikSchierboom) [inline concepts/functions/introduction.md:113]: This section can be removed as this exercise should not cover pointers (nor is it needed by the student to solve the exercise).

File: concepts/functions/introduction.md
Comment (ErikSchierboom) [inline concepts/functions/introduction.md:81]: This example uses multiple returns, which is something that is introduced in the `card-tricks` exercise. The dependencies here are tricky.

**Review by ErikSchierboom — COMMENTED** [review]
Thank you so much for doing this! I've roughly reviewed this, but decided to hold off on some reviewing more to first solve the dependency cycles.

> Fitting this concept/exercise into the existing hierarchy - I've done this somewhat, but if I add dependencies for everything that's relevant then I create cycles. Feedback welcomes from someone who has a more holistic view of the progression of concepts they think is right for the Go track.

@junedev could you perhaps take a look at how you envision the concept hierarchy? One immediate issue I could see was that this exercise depends on slices, but that the exercise teaching slices also teaches multiple returns. We could consider moving the multiple returns to this exercise? It's also important to note that we should decide if we want this to be an early exercise or not (as in: which exercises should depend on this)?

> The concept introduction and exercise introduction are nearly identical. Is this OK/appropriate? (Should I just be using .docs/introduction.md.tpl if so?)

This is perfectly ok. At some point the .tpl files will get good support, but we have only basic support for it now so I'd ignore it.

> .meta/design.md - this is my first attempt at one of these, I've read a few others but it's largely cargo cult.

It's perfectly fine.

> Should my exercise include forked_from? It's not clear to me from the docs what the contents of this field would be? (I.e. is it just same track exercises, or can you reference cross tracks etc)

Forked from is exclusively for cross-track references. So in this case it should be:

`"forked_from": ["javascript/lasagna-master"]

File: concepts/functions/introduction.md
Comment (bobtfish) [reply concepts/functions/introduction.md:113]: Damn, I knew I'd end up leaving something extraneous in the introduction, thanks for spotting!

File: concepts/functions/introduction.md
Comment (bobtfish) [reply concepts/functions/introduction.md:81]: The other option is to return a `struct` type - the `structs` concept only depends on `numbers` and `conditonals-if` so this doesn't seem unreasonable...

**Comment by bobtfish** [comment]
> Thank you so much for doing this! I've roughly reviewed this, but decided to hold off on some reviewing more to first solve the dependency cycles.

Sounds good - thanks for the rough cut that was really helpful.

Agree that we should sort the high level dependency issues and decide where this fits and what it can/can't depend on before we get into the minutiae.

File: concepts/functions/about.md
Comment (junedev) [reply concepts/functions/about.md:87]: No, that's not necessary. It is taught in the packages concept.

**Comment by junedev** [comment]
Disclaimer: I did not read the PR content yet, I am mostly responding to the question regarding the concept tree.

The functions concept should include explaining multiple return values, not as a separate concept just as part of the content. This is where it belongs. Content can probably be copied. Thus multiple return values can be safely used as part of the exercise here.

This exercise should have slices and structs as prerequisites. When you modify the scaleRecipe exercise to use a slice of structs instead of a map that should be enough to make it work. (Make sure not to use `range`, it not available yet. Just use a normal for loop.)

We can then later go ahead and remove the multiple return values concept from slices. The exercise would just need a minimal change, there is no conceptual dependency (unlike for maps, there you need multiple return values). Until then I don't think anyone will think this is bad track because we repeat a couple of lines of concept content. Everything that depended on multiple return values will then depend on functions later.

The general hierarchy on the end would be that the student learns the basic types and conditionals, that unlocks slices and structs, which unlocks functions which unlocks maps.

With regards to pointers I would go with the "learn the details later approach". Strictly speaking you would need pointers to understand how the complex  types are passed to the function. However currently pointers has maps as prerequisite and I would like to keep it a bit further down the tree. So instead of going into the details here I would just talk about how the passing works for "reference types" and primitives.

Does that help/ make sense?

**Comment by junedev** [comment]
I had a quick look at the exemplar file:
* I dont think the first task is needed here, it was about "not passing an argument" which does not exist in Go
* The second task should be changed to practice Gos version of a default parameter (an if statement). If the second argument is 0, use 2 as default.

File: concepts/functions/about.md
Comment (ErikSchierboom) [reply concepts/functions/about.md:87]: @bobtfish Could you remove the public/private section here (and in any other places in this PR)?

**Comment by ErikSchierboom** [comment]
@bobtfish Does @junedev's explanation help sort the dependencies issue?

File: concepts/functions/about.md
Comment (bobtfish) [reply concepts/functions/about.md:87]: This is not currently taught or mentioned anywhere in the packages concept or exercise that I can see, unless I'm missing it.

Does adding this need a ticket?

File: concepts/functions/introduction.md
Comment (bobtfish) [reply concepts/functions/introduction.md]: Per @junedev we're OK with multiple returns here, and will eventually remove them from `card-tricks` - resolving this.

**Comment by bobtfish** [comment]
> @bobtfish Does @junedev's explanation help sort the dependencies issue?

Yes! Thanks @junedev - I'll work on getting this into shape in the next day or two and let you know when it's ready for review.

**Comment by junedev** [comment]
Yes please, packages should teach how to import AND export things.
-------- Ursprüngliche Nachricht --------Von: Tomas Doran ***@***.***> Datum: 30.09.21  11:15  (GMT+01:00) An: exercism/go ***@***.***> Cc: June ***@***.***>, Mention ***@***.***> Betreff: Re: [exercism/go] Functions concept (#1720) 
@bobtfish commented on this pull request.



In concepts/functions/about.md:
> +## Public and Private Functions
+
+To allow a function to be called by [code in other packages][public-vs-private], the name of the function must start with a capital letter.

This is not currently taught or mentioned anywhere in the packages concept or exercise that I can see, unless I'm missing it.
Does adding this need a ticket?

File: concepts/functions/about.md
Comment (junedev) [reply concepts/functions/about.md:87]: Yes please, packages should teach how to import AND export things.

**Comment by ErikSchierboom** [comment]
@junedev I've opened an issue: https://github.com/exercism/go/issues/1728

**Comment by junedev** [comment]
@bobtfish Another thought regarding scaleRecipe. Instead of making it a slice of structs you could also just have a slice of floats an say "each item represents the amount of some ingredient ...". The task is about not modifying the argument so a slice will do to make this point.

**Comment by junedev** [comment]
@bobtfish Setting this to draft for now. Mark as "ready for review" again whenever you are ready.

**Comment by bobtfish** [comment]
> Make sure not to use `range`, it not available yet. Just use a normal for loop.

I'm unsure about this - it's totally possible to do this, but the exercise is a lot nicer to solve if you can use `range`, and there isn't anything in the dependencies at the moment which stops this from depending on `range-iteration`.

Why do we imagine that this has to be before we cover `range-iteration`?

**Comment by junedev** [comment]
> Why do we imagine that this has to be before we cover range-iteration?

Range iteration need to show how to iterate over maps so maps need to be known to the student, maps need multiple return values from here so we can't move maps before functions.

**Comment by bobtfish** [comment]
> > Why do we imagine that this has to be before we cover range-iteration?
> 
> Range iteration need to show how to iterate over maps so maps need to be known to the student, maps need multiple return values from here so we can't move maps before functions.

Aha, got it, thanks for the explanation. I'll take `range-iteration` out later tonight.

File: exercises/concept/lasagna-master/.docs/instructions.md
Comment (junedev) [inline exercises/concept/lasagna-master/.docs/instructions.md:17]:
```suggestion
If the average preperation time is passed as `0` (the default initial value for an `int`), then the default value of `2` should be used.
```

File: exercises/concept/lasagna-master/.meta/design.md
Comment (junedev) [inline exercises/concept/lasagna-master/.meta/design.md:37]: I think you removed this requirement.
```suggestion
```

File: exercises/concept/lasagna-master/.meta/exemplar.go
Comment (junedev) [inline exercises/concept/lasagna-master/.meta/exemplar.go:3]: All function comments need to conform to the Go spec (see concept comments).

**Review by junedev — COMMENTED** [review]
Well done overall! I really like the examples you used in the concepts.

Some more small things need to be fixed before this can be merged.

**Comment by bobtfish** [comment]
> Well done overall! I really like the examples you used in the concepts.

Thanks!

> Some more small things need to be fixed before this can be merged and the merge conflict needs to be resolved.

All fixed up. I wasn't 100% sure what you were suggesting around the `PrintHelloName` example, but I have had a go at amending it.

I've also rebased and squashed as it was getting unwieldy - no point committing adding then removing then re-adding the same thing , and that also allowed me to fix the merge conflict so it'll merge cleanly without a load of back-and-forth merges to make future history exploration easy. (I'm not sure if there's a house style/recommendation about merges/rebases for PRs in progress)

**Review by junedev — APPROVED** [review]

**Pull request merged**
